specification/messaging/messaging-metrics.md

```diff
@@ -0,0 +1,257 @@
+| `messaging.todo.status` | string | The status of the operation: TODO | `ok`; `error`; `timeout` | Required |
+| `messaging.destination.name` | string | The message destination name [1] | `MyQueue`; `MyTopic` | Conditionally Required: [2] |
+| `messaging.destination.template` | string | Low cardinality representation of the messaging destination name [3] | `/customers/{customerId}` | Conditionally Required: if available. |
+| [`network.protocol.name`](../general/general-attributes.md) | string | [OSI Application Layer](https://osi-model.com/application-layer/) or non-OSI equivalent. The value SHOULD be normalized to lowercase. | `amqp`; `mqtt` | Recommended |
+| [`network.protocol.version`](../general/general-attributes.md) | string | Version of the application layer protocol used. See note below. [4] | `3.1.1` | Recommended |
+| [`server.address`](../general/general-attributes.md) | string | Logical server hostname, matches server FQDN if available, and IP or socket address if FQDN is not known. | `example.com` | Conditionally Required: If available. |
+| [`server.port`](../general/general-attributes.md) | int | Logical server port number | `80`; `8080`; `443` | Conditionally Required: If available. |
+| [`server.socket.address`](../general/general-attributes.md) | string | Physical server IP address or Unix socket address. If set from the client, should simply use the socket's peer address, and not attempt to find any actual server IP (i.e., if set from client, this may represent some proxy server instead of the logical server). | `10.5.3.2` | Recommended: If different than `server.address`. |
+
+**[1]:** Destination name SHOULD uniquely identify a specific queue, topic or other entity within the broker. If
+the broker does not have such notion, the destination name SHOULD uniquely identify the broker.
+
+**[2]:** if and only if `messaging.destination.name` is known to have low cardinality. Otherwise, `messaging.destination.template` MAY be populated.
+
+**[3]:** Destination names could be constructed from templates. An example would be a destination name involving a user name or product id. Although the destination name in this case is of high cardinality, the underlying template is of low cardinality and can be effectively used for grouping and aggregation.
+
+**[4]:** `network.protocol.version` refers to the version of the protocol used and might be different from the protocol client's version. If the HTTP client used has a version of `0.27.2`, but sends HTTP version `1.1`, this attribute should be set to `1.1`.
+<!-- endsemconv -->
+
+## Metric: `messaging.received_messages.count`
+
+This metric is recommended when messaging system supports batch receive. For the systems that can only receive one message, message count can be derived from `messaging.receive.duration` or `messaging.deliver.duration` histogram.
```

Comment:
Wouldn't it be better to recommend these metrics for all cases, not just for batch receive scenarios? Otherwise, I imagine it will be hard to interpret those metrics (e. g. when to count the histogram, and when to look at the counters).

Also, there's the case when both batches and single message might be received.

Reply:
I updated wording to reflect that this condition applies to the messaging system (e.g. rabbitmq does not support batching and rabbitmq instrumentation would never need to report counts). Let me know if new wording makes more sense or if you still see a problem in it.

Reply:
The wording makes sense to me. However, a concern I have with a logic like this is, that we loose the ability to interpret the metrics in a messaging-system independent way.

The way proposed here seems to be the most efficient for telemetry producers, however I imagine it to be hard for telemetry consumers to know when the `*.duration` metrics have the additional semantic of also representing the count of messages. This will then differ from system to system.

Reply:
Discussion in the WG:
* Counters are required for systems that support batching.
* For other systems, counters are opt-in.

Reply:
addressed, PTAL